### PR TITLE
feat: add thread safety type WebviewMut

### PR DESCRIPTION
**Pull request by wusyong**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] New Binding Issue #___
- [ ] Code style update
- [x] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [x] Yes.
- [ ] No

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch and _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
This PR adds some addition in the library to make users able to call webview methods with thread safety manner for sure.
- `Webview` instance remains the same which has `!Send`/`!Sync`.
- **`WebviewMut` is introduced which can be created by `as_mut` method from `Webview`**. This is a NewType pattern for users calling API across threads. Note that only chosen methods are implemented in this type. Some calls like `run` should be definitely called by main thread. Here are the methods provided in WebviewMut:
  - `terminate`
  - `get_window`
  - `navigate`
  - `eval`
  - `bind`
  - `r#return`
- `WebviewRef` is not introduced since the only call it can provide is `get_window` which is pretty useless imo. Let me know if anyone still thinks this is necessary.
- Remove destructors `destroy` since it shouldn't expose to public. The drop trait should handle itself.
- Add a example `run_and_terminate` to show how users could do other than main thread.

### CODE REVIEW

File: src/webview.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        unsafe { sys::webview_get_window(*self.0) as *mut Window }
+    }
+
+    pub fn navigate(&mut self, url: &str) {
+        let c_url = CString::new(url).expect("No nul bytes in parameter url");
+        unsafe { sys::webview_navigate(*self.0, c_url.as_ptr()) }
+    }
+
+    pub fn init(&mut self, js: &str) {
+        let c_js = CString::new(js).expect("No nul bytes in parameter js");
```
Comment (nothingismagick) [inline src/webview.rs]:
```suggestion
        let c_js = CString::new(js).expect("No null bytes in parameter js");
```

File: src/webview.rs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    pub fn set_size(&mut self, width: i32, height: i32, hints: SizeHint) {
+        unsafe { sys::webview_set_size(*self.0, width, height, hints as i32) }
+    }
+
+    pub fn get_window(&self) -> *mut Window {
+        unsafe { sys::webview_get_window(*self.0) as *mut Window }
+    }
+
+    pub fn navigate(&mut self, url: &str) {
+        let c_url = CString::new(url).expect("No nul bytes in parameter url");
```
Comment (nothingismagick) [inline src/webview.rs]:
```suggestion
        let c_url = CString::new(url).expect("No null bytes in parameter url");
```

File: src/webview.rs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        unsafe { sys::webview_terminate(*self.0) }
+    }
+
+    pub fn as_mut(&mut self) -> WebviewMut {
+        WebviewMut(Arc::downgrade(&self.0))
+    }
+
+    // TODO Window instance
+    pub fn set_title(&mut self, title: &str) {
+        let c_title = CString::new(title).expect("No nul bytes in parameter title");
```
Comment (nothingismagick) [inline src/webview.rs]:
```suggestion
        let c_title = CString::new(title).expect("No null bytes in parameter title");
```

File: src/webview.rs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                *self.0,
+                c_name.as_ptr(),
+                Some(callback::<F>),
+                closure as *mut _,
+            )
+        }
+    }
+
+    pub fn r#return(&self, seq: &str, status: c_int, result: &str) {
+        let c_seq = CString::new(seq).expect("No nul bytes in parameter seq");
```
Comment (nothingismagick) [inline src/webview.rs]:
```suggestion
        let c_seq = CString::new(seq).expect("No null bytes in parameter seq");
```

File: src/webview.rs [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        unsafe { sys::webview_navigate(*self.0, c_url.as_ptr()) }
+    }
+
+    pub fn init(&mut self, js: &str) {
+        let c_js = CString::new(js).expect("No nul bytes in parameter js");
+        unsafe { sys::webview_init(*self.0, c_js.as_ptr()) }
+    }
+
+    pub fn eval(&mut self, js: &str) {
+        let c_js = CString::new(js).expect("No nul bytes in parameter js");
```
Comment (nothingismagick) [inline src/webview.rs]:
```suggestion
        let c_js = CString::new(js).expect("No null bytes in parameter js");
```

File: src/webview.rs [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+                c_name.as_ptr(),
+                Some(callback::<F>),
+                closure as *mut _,
+            )
+        }
+    }
+
+    pub fn r#return(&self, seq: &str, status: c_int, result: &str) {
+        let c_seq = CString::new(seq).expect("No nul bytes in parameter seq");
+        let c_result = CString::new(result).expect("No nul bytes in parameter result");
```
Comment (nothingismagick) [inline src/webview.rs]:
```suggestion
        let c_result = CString::new(result).expect("No null bytes in parameter result");
```

File: src/webview.rs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    pub fn get_window(&self) -> Result<*mut Window, Error> {
+        let webview = self.0.upgrade().ok_or(Error::WebviewNull)?;
+        Ok(unsafe { sys::webview_get_window(*webview) as *mut Window })
+    }
+
+    pub fn navigate(&mut self, url: &str) -> Result<(), Error> {
+        let webview = self.0.upgrade().ok_or(Error::WebviewNull)?;
+        let c_url = CString::new(url).expect("No nul bytes in parameter url");
```
Comment (nothingismagick) [inline src/webview.rs]:
```suggestion
        let c_url = CString::new(url).expect("No null bytes in parameter url");
```

File: src/webview.rs [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        unsafe { sys::webview_navigate(*webview, c_url.as_ptr()) }
+        Ok(())
+    }
+
+    pub fn eval(&mut self, js: &str) -> Result<(), Error> {
+        let webview = self.0.upgrade().ok_or(Error::WebviewNull)?;
+        let c_js = CString::new(js).expect("No nul bytes in parameter js");
```
Comment (nothingismagick) [inline src/webview.rs]:
```suggestion
        let c_js = CString::new(js).expect("No null bytes in parameter js");
```

File: src/webview.rs [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+                Some(callback::<F>),
+                closure as *mut _,
+            )
+        }
+        Ok(())
+    }
+
+    pub fn r#return(&self, seq: &str, status: c_int, result: &str) -> Result<(), Error> {
+        let webview = self.0.upgrade().ok_or(Error::WebviewNull)?;
+        let c_seq = CString::new(seq).expect("No nul bytes in parameter seq");
```
Comment (nothingismagick) [inline src/webview.rs]:
```suggestion
        let c_seq = CString::new(seq).expect("No null bytes in parameter seq");
```

File: src/webview.rs [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        let c_result = CString::new(result).expect("No nul bytes in parameter result");
```
Comment (nothingismagick) [inline src/webview.rs]:
```suggestion
        let c_result = CString::new(result).expect("No null bytes in parameter result");
```

**Review by nothingismagick — COMMENTED** [review]
nul again :D

**Review by nothingismagick — APPROVED** [review]

**Comment by nothingismagick** [comment]
`get_window` -> can it help resolve my concerns about the scope of the callee as mentioned here:
https://github.com/tauri-apps/tauri/issues/814

**Comment by wusyong** [comment]
Window API is still not defined in webview yet. But I guess it can provide the possibility. Might need to dive into it deeper to figure out.

**Pull request merged**